application/common/models/Password.php

```diff
@@ -236,4 +236,40 @@ public function scenarios(): array
 
         return $scenarios;
     }
+
+    public function extendHibpCheckAfter(): void
+    {
+        $this->setScenario(self::SCENARIO_UPDATE_METADATA);
+        $this->check_hibp_after = MySqlDateTime::relativeTime(\Yii::$app->params['hibpCheckInterval']);
+        if (!$this->save()) {
+            \Yii::warning([
+                'action' => 'extend hibp check after',
+                'employee_id' => $this->employee_id,
+                'message' => 'unable to update check_hibp_after',
+                'errors' => $this->getFirstErrors(),
+            ]);
+        }
+    }
+
+    /**
+     * Mark password as pwned by:
+     *  - set hibp_is_pwned to yes
+     *  - set expiration to now
+     *  - set graceperiod based on config
+     */
+    public function markPwned(): void
+    {
+        $this->setScenario(self::SCENARIO_UPDATE_METADATA);
+        $this->hibp_is_pwned = 'yes';
+        $this->expires_on = MySqlDateTime::relativeTime('-5 minutes');
+        $this->grace_period_ends_on = MySqlDateTime::relativeTime(\Yii::$app->params['hibpGracePeriod']);
```

Comment:
I think this will not work as intended for a user with MFA options. In such a case, the value returned to the API consumer is extended from the database value.

Reply:
oh good point. So should I set `expires_on` to something like 20 years ago to be safe?

Reply:
I don't think that would work. It's the grace period that matters, I think, and unless you check for the presence of a valid MFA, you won't know how to set `grace_period_ends_on`. Thinking right now there are a few options, (1) check for MFA and set the grace period accordingly, or (2) revise how the grace period is implemented so the stored grace period is actually what is returned to the consumer, or (3) leave it as it is knowing that a user with MFA is relatively safe even with a pwned password.

Reply:
Thanks @Schparky that's good looking out! What about this solution I thought of?

```php
    /**
     * If password is pwned, return actual expires_on.
     * Otherwise calculate expires_on date based on if user has MFA configured
     * @return string
     */
    public function getExpiresOn()
    {
        return $this->hibp_is_pwned == 'yes' ? $this->expires_on : $this->getMfaExtendedDate($this->expires_on);
    }

    /**
     * If password is pwned, return actual grace_period_ends_on.
     * Otherwise calculate grace_period_ends_on based on if user has MFA configured
     * @return string
     */
    public function getGracePeriodEndsOn()
    {
        return $this->hibp_is_pwned == 'yes' ? $this->grace_period_ends_on : $this->getMfaExtendedDate($this->grace_period_ends_on);
    }
```

Since these functions already existed inside `Password` it seemed reasonable to take into account `$this->hibp_is_pwned`.

Reply:
Looks like a perfect solution!